kiwmi/desktop/xdg_shell.c

```diff
@@ -22,6 +23,158 @@
 #include "input/seat.h"
 #include "server.h"
 
+static struct kiwmi_view_child *view_child_popup_create(
+    struct kiwmi_view_child *parent,
+    struct kiwmi_view *view,
+    struct wlr_xdg_popup *wlr_popup);
+
+static void
+popup_new_popup_notify(struct wl_listener *listener, void *data)
+{
+    struct kiwmi_view_child *popup =
+        wl_container_of(listener, popup, new_popup);
+    struct wlr_xdg_popup *wlr_popup = data;
+    view_child_popup_create(popup, popup->view, wlr_popup);
+}
+
+static void
+popup_extension_destroy_notify(struct wl_listener *listener, void *UNUSED(data))
+{
+    struct kiwmi_view_child *popup =
+        wl_container_of(listener, popup, extension_destroy);
+    view_child_destroy(popup);
+}
+
+static void
+popup_unconstrain(struct kiwmi_view_child *popup)
+{
+    if (popup->type != KIWMI_VIEW_CHILD_XDG_POPUP) {
+        wlr_log(WLR_ERROR, "Expected an xdg_popup kiwmi_view_child");
+        return;
+    }
+
+    struct kiwmi_view *view         = popup->view;
+    struct wlr_xdg_popup *wlr_popup = popup->wlr_xdg_popup;
+
+    double popup_sx;
+    double popup_sy;
+    wlr_xdg_popup_get_position(wlr_popup, &popup_sx, &popup_sy);
+
+    // TODO: possibly ignore the popup's own position. When moving a view from
+    // the very right of an output to the very left, popups may end up on an
+    // output to the left of the view's output.
+    struct wlr_output *output = wlr_output_layout_output_at(
+        view->desktop->output_layout, view->x + popup_sx, view->y + popup_sy);
```

Comment:
What would you say is better? Normally I would expect ignoring the popup’s position to be okay, but it can have unexpected results when a view spans multiple outputs.

Reply:
I think not taking the output into account makes sense here.

I'm not sure what's the best approach here is, do you know what the other compositors do?

Reply:
I think not taking the output into account makes sense here.

I'm not sure what's the best approach here is, do you know what the other compositors do?

Reply:
Another possibility is to give control to the user, e.g. via `view:unconstrain_rect(lx, ly, w, h)` (or `vx, vy`) defaulting to the view’s rect.

That way this function doesn’t have to be so complicated and the user can always do what they wish.

Reply:
Hmm, that is also not a good idea.

If we use `vx, vy`, it’s rather illogical, because a user is most likely to want to use an output’s rect, which is in `lx, ly`, and they have to update the unconstrain rect every time they move a view.

If we use `lx, ly`, it can become unpredictable. A user is likely to expect the unconstrain rect to be applied automatically when the view moves, but if they want to update the rect at the same time, `popup_unconstrain` is called twice and the popup can move even if it wouldn’t really have to (as far as the user cares).